### PR TITLE
Jk/cumulus 3072

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Updates PUT endpoint to utilize RESTful PUT logic

Testers should:

- Do a thorough code review
- Deploy and evaluate API responses for expected results across all datastores

Addresses [CUMULUS-3072:](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3072)

- **CUMULUS-3072**
  - Removed original PUT granule endpoint logic (in favor of utilizing new PATCH
    endpoint introduced in CUMULUS-3071)
  - Updated PUT granule endpoint to expected RESTful behavior:
    - PUT will now overwrite all non-provided fields as either non-defined or
      defaults, removing existing related database records (e.g. files,
      granule-execution linkages ) as appropriate.
    - PUT will continue to overwrite fields that are provided in the payload,
      excepting collectionId and granuleId which cannot be modified.
    - PUT will create a new granule record if one does not already exist
    - Like PATCH, the execution field is additive only - executions, once
      associated with a granule record cannot be unassociated via the granule endpoint.

  - Added `replaceGranule` to `@cumulus/api-client/granules` to add usage of the
    updated RESTful PUT logic

  - Fixed issue introduced in CUMULUS-3070 where new granules incorrectly write
    a value for `files` as `[]` to elasticsearch instead of undefined in cases
    where `[]` is specified in the new granule.
  - Fixed issue introduced in CUMULUS-3070 where DynamoDB granules with a value
   `files` as `[]` when the granule does *not* have the files value set as
   mutable (e.g. in a `running` state) from a framework message write *and*
   files was not previously defined will write `[]` instead of leaving the value
   undefined.

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

### CODE REVIEW

File: CHANGELOG.md
Comment (Jkovarik) [inline CHANGELOG.md:95]: These logic changes were required for this PR to continue work, however these changes may be ported to main before this PR merges.

File: packages/api/models/base.js
Comment (jennyhliu) [inline packages/api/models/base.js:479]: I don't quite understand the difference from previous implementation

File: packages/api/tests/endpoints/test-granules.js
Comment (jennyhliu) [inline packages/api/tests/endpoints/test-granules.js:2119]: Should we update the deleted `PUT` tests to `PATCH` tests?

File: packages/api/tests/endpoints/test-granules.js
Comment (jennyhliu) [inline packages/api/tests/endpoints/test-granules.js:3459]: TODO?

File: packages/api/tests/endpoints/test-granules.js
Comment (jennyhliu) [inline packages/api/tests/endpoints/test-granules.js:3504]: If `files` is deleted from data stores, should the actual files get deleted from s3?

**Review by jennyhliu — CHANGES_REQUESTED** [review]

File: packages/api-client/src/granules.ts
Comment (Jkovarik) [reply packages/api-client/src/granules.ts:452]: Ah yes, good catch, I'd forgotten we had these.     👍🏻

File: packages/api/tests/endpoints/test-granules.js
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-granules.js:3504]: No, at least not in *this* PR - none of our other granules endpoint APIs do that *except* delete granules because they were implemented under different feature requests (e.g. delete granule was implemented before modification/insertions was possible as a file/granule management feature whereas this API was built with expectations of database management only, like the workflow granule.files (they're not removed on duplicate ingest). 

I'd be happy to talk more about that in detail, I feel pretty strongly that the API should be consistent, and we should have a discussion about reingest and excess files.

File: packages/api/tests/endpoints/test-granules.js
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-granules.js:2119]: Looking at this, it looks like a possible merge issue 😦

File: packages/api/models/base.js
Comment (Jkovarik) [reply packages/api/models/base.js:479]: @jennyhliu  see the unit test updates here https://github.com/nasa/cumulus/pull/3189/files#diff-5db7829f5b5faa6a3ce55a161353fbcec8489ae540fad4fb2ada5194415c8432R1156

The issue is that files is updated ONLY if it doesn't already exist *OR* if it's in the mutable list.  

Meaning that in the prior implementation if a granule is: 

{ granuleId: someId, collectionId: someId, files: [], ....} 

and it's *not* in the mutable field names, *but* the value isn't set, whatever is in `item.files` gets written.   

In the case of values, this is intended.   In the case of null, it's handled, and in the case of it not being defined, nothing gets written as expected. 

`[]` is a special case, however as we want it to remain undefined.   This update decouples (and refactors):

`if (this.parseEmptyFilesArrayAsNull)` -- If we care about the files key being parsed as null
`(fieldName === 'files' && mutableFieldNames.includes(fieldName) && isEqual(value, []))` -- If the field name is files, *and* it's supposed to be mutated *and* it's the special value:

set the value to be deleted and add the expression attribute for the query, but don't set anything.       In this case, if the field name is not set to be mutable, it doesn't do anything.   Critically that means that *it may still write* if the value isn't present. 

Instead we say 

`if (this.parseEmptyFilesArrayAsNull && fieldName === 'files')` - If the field name matches and it's configured to handle it (refactor)
`if (isEqual(value, [])) {` -- if it's the special value, take action, if not continue as normal 
`if (mutableFieldNames.includes(fieldName))` { --  if it's a mutable field, set it to be deleted.  

 IF *not* just return, which does not add it to the list of updates, which means it won't modify it if it exists, and will not insert anything if it does not. 
 
 I'd be happy to demo through that if that doesn't make sense/feels incorrect/etc.

File: packages/api-client/src/granules.ts
Comment (Jkovarik) [reply packages/api-client/src/granules.ts:452]: Check out 0980bb79cc - it adds a new granule, then validates a PUT with nothing but the minimum values does what's expected.

File: packages/api/tests/endpoints/test-granules.js
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-granules.js:2119]: re-added these, updated to the correct endpoints and fixed a few other things in the units: 

59239b9fc3
ed3f2ea666

**Comment by Jkovarik** [comment]
@jennyhliu  I think I've answered your questions/addressed comments.   Let me know what you think, and many thanks for the review!

File: example/spec/parallel/testAPI/granuleSpec.js
Comment (jennyhliu) [inline example/spec/parallel/testAPI/granuleSpec.js:390]: Should we replace a granule which has execution associated?

**Comment by jennyhliu** [comment]
Verified that PUT endpoint works as expected.  I will approve when it's time to merge to master.

**Comment by Jkovarik** [comment]
@jennyhliu  I've merged #3257  with the API versioning updates to the branch and it should be good to go.

**Comment by Jkovarik** [comment]
@jennyhliu  that's an excellent point.... I *think* given it uses @cumulus/api we should just be able to build it with the changes in #3257  and be fine -- to my knowledge dash doesn't make granule *modification* queries, the only impact here is that we've conflated the bulk operations stuff with granule PUT/PATCH operations (which we should sort).

**Comment by Jkovarik** [comment]
@jennyhliu  based on the convos during standup, I think this ticket shouldn't be blocked on the dashboard ticket, however I *am* going to block this PR on cutting the v14.1 release branch.     Please let me know if you feel there are any other outstanding criteria on this PR re: approval.

**Comment by jennyhliu** [comment]
Are you going to add a dashboard ticket or PR?

**Comment by Jkovarik** [comment]
@jennyhliu  the intent is to update the dashboard, yes, so a PR.      It's status as a separate ticket is being discussed, however the intent is to merge this once v15 with the node updates is out, and the next release (v16) would be blocked on the dashboard updates.

**Review by jennyhliu — APPROVED** [review]
Nice work!

File: packages/api/endpoints/granules.js
Comment (npauzenga) [inline packages/api/endpoints/granules.js:273]: @Jkovarik it looks like `put` doesn't support actions. Is that right and is PATCH the endpoint that we'll use to (e.g.) applyWorkflow? It seems like that could be right but I wanted to verify because that would be a breaking change I think as PUT currently supports actions.

Comment (Jkovarik) [reply packages/api/endpoints/granules.js:273]: @npauzenga  yes, that's intended.  I think that's correct (insofar as having actions in a PUT/PATCH endpoint *at all* is correct) as those actions more correctly align with PATCH (modification of but not necessarily complete replacement of the record).

I'm open to other thoughts on the approach if you feel strongly.

**Comment by Jkovarik** [comment]
resetting this to in-progress, PR will be applied on top of the P3 release branch/main, so re-review will be required.

File: packages/api/endpoints/granules.js
Comment (npauzenga) [inline packages/api/endpoints/granules.js:249]: I think this comment is resolved?

File: packages/api/endpoints/granules.js
Comment (npauzenga) [inline packages/api/endpoints/granules.js:272]: TODO?

File: packages/api/lib/request.js
Comment (npauzenga) [inline packages/api/lib/request.js:57]: TODO?

File: packages/api/tests/endpoints/test-granules.js
Comment (npauzenga) [inline packages/api/tests/endpoints/test-granules.js:3598]: Can probably remove the log

File: packages/api/tests/endpoints/test-granules.js
Comment (npauzenga) [inline packages/api/tests/endpoints/test-granules.js:3692]: This relates to an early comment. So the preference is to use the payload for the granuleId and collectionId? If those aren't supplied, use the URI? Could you briefly explain the thinking here just for my understanding?

**Review by npauzenga — CHANGES_REQUESTED** [review]
This is looking good @Jkovarik. Re-reviewed the changes now that they're targeting main. I had a couple minor comments and one area where I could use clarification (ID params from payload vs. URI).

File: packages/api/endpoints/granules.js
Comment (Jkovarik) [reply packages/api/endpoints/granules.js:272]: Resolved - out of scope/separate ticket/still need.

File: packages/api/tests/endpoints/test-granules.js
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-granules.js:3692]: Discussed in huddle - the idea is to add values to the payload from the URI if they're missing as convenience to users.      Will update code and resubmit for review to do this for granuleId as well.

**Review by npauzenga — APPROVED** [review]
Looks great to me @Jkovarik!

**Pull request merged**
